SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -627,35 +710,209 @@ void ASpatialDebugger::DrawDebug(UCanvas* Canvas, APlayerController* /* Controll
-				if (LocalPlayerController.IsValid())
+			// Draw tags above selected actors
+			for (TWeakObjectPtr<AActor> SelectedActor : SelectedActors)
+			{
+				if (SelectedActor.IsValid())
 				{
-					SCOPE_CYCLE_COUNTER(STAT_Projection);
-					UGameplayStatics::ProjectWorldToScreen(LocalPlayerController.Get(), ActorLocation + WorldSpaceActorTagOffset,
-														   ScreenLocation, false);
+					if (const Worker_EntityId_Key* HitEntityId = EntityActorMapping.FindKey(SelectedActor))
+					{
+						FVector PlayerLocation = GetLocalPawnLocation();
+
+						if (SelectedActor != nullptr)
+						{
+							FVector2D ScreenLocation = ProjectActorToScreen(SelectedActor, PlayerLocation);
+							if (!ScreenLocation.IsZero())
+							{
+								DrawTag(Canvas, ScreenLocation, *HitEntityId, SelectedActor->GetName(), true /*bCentre*/);
+							}
+						}
+					}
 				}
+			}
+		}
+	}
+}
 
-				if (ScreenLocation.IsZero())
+void ASpatialDebugger::RevertHoverMaterials()
+{
+	if (HoverActor != nullptr)
+	{
+		// Revert materials on previous actor
+		for (int i = 0; i < ActorMeshComponents.Num(); i++)
+		{
+			UActorComponent* ActorMeshComponent = ActorMeshComponents[i];
+			UMeshComponent* ActorStaticMeshComponent = Cast<UMeshComponent>(ActorMeshComponent);
+			ActorStaticMeshComponent->SetMaterial(0, ActorMeshMaterials[i]);
+		}
+
+		// Clear previous materials
+		ActorMeshMaterials.Empty();
+		ActorMeshComponents.Empty();
+
+		HoverActor = nullptr;
+	}
+}
+
+TWeakObjectPtr<AActor> ASpatialDebugger::GetActorAtPosition(FVector2D& NewMousePosition)
+{
+	if (LocalPlayerController == nullptr)
+	{
+		return nullptr;
+	}
+	else if (NewMousePosition != MousePosition)
+	{
+		// Mouse has moved so raycast to find actors currently under the mouse cursor
+		MousePosition = NewMousePosition;
+
+		FVector WorldLocation;
+		FVector WorldRotation;
+		LocalPlayerController->DeprojectScreenPositionToWorld(NewMousePosition.X, NewMousePosition.Y, WorldLocation,
+															  WorldRotation); // Mouse cursor position
+		FVector StartTrace = WorldLocation;
+		FVector EndTrace = StartTrace + WorldRotation * MaxRange;
+
+		HitActors.Empty();
+
+		TArray<FHitResult> HitResults;
+		bool bHit = GetWorld()->LineTraceMultiByObjectType(HitResults, StartTrace, EndTrace, CollisionObjectParams);
+		if (bHit)
+		{
+			// When the raycast hits an actor then it is highlighted, whilst the actor remains under the crosshair. If there are multiple
+			// hit results, the user can select the next by using the mouse scroll wheel
+			for (const FHitResult& HitResult : HitResults)
+			{
+				const TWeakObjectPtr<AActor> HitActor = HitResult.GetActor();
+
+				if (HitActors.Contains(HitActor))
 				{
+					// The hit results may include the same actor multiple times so just ignore duplicates
 					continue;
 				}
 
-				DrawTag(Canvas, ScreenLocation, EntityId, Actor->GetName(), true /*bCentre*/);
+				if (const Worker_EntityId_Key* HitEntityId = EntityActorMapping.FindKey(HitResult.GetActor()))
```

Comment:
Why do we we get the entity ID and screen position here?

Reply:
We don't want to add actors to the list of hit actors unless they have a valid entity id or screen position. As when we scroll through the actors we only want to highlight ones that we can show a tag for.

Reply:
Aha makes sense - maybe add a comment with that explanation here?

Reply:
Good point. I've added a comment to the code.